examples/langchain_rag/app.py

```diff
@@ -33,127 +55,136 @@ class CustomCallbackHandler(AsyncCallbackHandler):
-        run_id: UUID,
-        parent_run_id: Optional[UUID] = None,
-        tags: Optional[List[str]] = None,
-        **kwargs: Any,
-    ) -> None:
+    async def on_llm_new_token(self, token: str, *args, **kwargs) -> None:
         self._choice.append_content(token)
 
 
+def sanitize_namespace(namespace: str) -> str:
+    return "".join(c if c.isalnum() or c in "._-/" else "-" for c in namespace)
+
+
+def get_last_attachment_url(messages: List[Message]) -> str:
+    for message in reversed(messages):
+        if (
+            message.custom_content is not None
+            and message.custom_content.attachments is not None
+        ):
+            attachments = message.custom_content.attachments
+
+            if attachments == []:
+                continue
+
+            if len(attachments) != 1:
+                msg = "Only one attachment per message is supported"
+                raise DIALException(
+                    status_code=422,
+                    message=msg,
+                    display_message=msg,
+                )
+
+            attachment = attachments[0]
+
+            url = attachment.url
+            if url is None:
+                msg = "Attachment is expected to be provided via a URL"
+                raise DIALException(
+                    status_code=422,
+                    message=msg,
+                    display_message=msg,
+                )
+
+            return url
+
+    msg = "No attachment was found"
+    raise DIALException(
+        status_code=422,
+        message=msg,
+        display_message=msg,
+    )
+
+
 class SimpleRAGApplication(ChatCompletion):
     async def chat_completion(
         self, request: Request, response: Response
     ) -> None:
         collection_name = str(uuid4())
 
         with response.create_single_choice() as choice:
-            first_user_message = request.messages[0].content or ""
-            if first_user_message.endswith(".pdf"):
-                loader = PyPDFLoader(first_user_message)
-            else:
-                loader = WebBaseLoader(first_user_message)
+            message = request.messages[-1]
+            user_query = message.content or ""
 
-            openai_embedding = OpenAIEmbeddings(
-                openai_api_base=DIAL_URL,
-                openai_api_key=request.api_key,
-                openai_api_version="2023-03-15-preview",
-                openai_api_type="azure",
-                max_retries=15,
-                headers=(
-                    {} if not request.jwt else {"Authorization": request.jwt}
-                ),
-            )
-
-            embeddings = CacheBackedEmbeddings.from_bytes_store(
-                openai_embedding,
-                embedding_store,
-                namespace=openai_embedding.model,
-            )
+            file_url = get_last_attachment_url(request.messages)
+            file_abs_url = urljoin(f"{DIAL_URL}/v1/", file_url)
```

Comment:
Previously, the application worked with links to websites. Why do we want to change this?

Reply:
It still works with links to websites.